drivers/can/can_xlnx.c

```diff
@@ -0,0 +1,1329 @@
+		 */
+		sys_write32(0x0, base_addr + CAN_XLNX_SRR_OFFSET);
+
+		/* Wait for SR[CONFIG] */
+		do {
+			sr = sys_read32(base_addr + CAN_XLNX_SR_OFFSET);
+		} while (!(sr & CAN_XLNX_SR_CONFIG_MODE));
+
+		/*
+		 * Entering config mode clears the error counters
+		 * -> clear the stored error counters.
+		 */
+		dev_data->rx_errors = 0;
+		dev_data->tx_errors = 0;
+
+		LOG_DBG("%s is now in configuration mode", dev->name);
+	}
+}
+
+/**
+ * @brief Bus state transition function
+ * @internal
+ *
+ * Performs a bus state transition, if new state != current state.
+ * This function is called from the ISR for handling of the bus off
+ * state, can_xlnx_handle_errors for error active / error warning /
+ * error passive state handling based on the error counters, and
+ * can_xlnx_set_mode whenever an operational mode other than con-
+ * figuration mode is established. If a state change callback has
+ * been registered with the current driver instance, it will be
+ * executed.
+ *
+ * @param dev Pointer to the CAN controller's device.
+ * @param state Potentially new bus state detected in one of the
+ *              contexts described above.
+ */
+static void can_xlnx_state_update(const struct device *dev,
+				  enum can_state state)
+{
+	struct can_xlnx_dev_data *dev_data = dev->data;
+	struct can_bus_err_cnt error_counters;
+
+	if (state != dev_data->state) {
+		dev_data->state = state;
+		LOG_DBG("%s is now in bus state %s", dev->name,
+			(state == CAN_ERROR_ACTIVE) ? "CAN_ERROR_ACTIVE" :
+			(state == CAN_ERROR_PASSIVE) ? "CAN_ERROR_PASSIVE" :
+			(state == CAN_ERROR_WARNING) ? "CAN_ERROR_WARNING" :
+			"CAN_BUS_OFF");
+
+		/*
+		 * When transitioning to bus off state:
+		 * -> Set separate marker in the driver instance's data as this
+		 * state can not be expressed in RX/TX error counter values. We
+		 * know that bus off is the new state as this function is only
+		 * ever called with this new state based on the evaluation of
+		 * the corresponding bit in the Interrupt Status Register.
+		 * While the bus off state could also be obtained from SR[ESTAT]
+		 * = 0x2 (comp. Zynq-7000 TRM, register SR description, Appendix
+		 * B.5, p. 804), a local variable in the device data is probably
+		 * preferrable compared to reading a register every time the bus
+		 * state is queried.
+		 * -> Clear the local error counters, as this happens auto-
+		 * matically in hardware when entering the bus off state.
+		 */
+		if (state != CAN_BUS_OFF) {
+			dev_data->bus_off = 0;
+		} else {
+			dev_data->bus_off = 1;
+			dev_data->rx_errors = 0;
+			dev_data->tx_errors = 0;
+		}
+
+		/* Execute state change callback if registered */
+		if (dev_data->state_chg_cb != NULL) {
+			LOG_DBG("%s calling state change callback function",
+				dev->name);
+
+			error_counters.rx_err_cnt = dev_data->rx_errors;
+			error_counters.tx_err_cnt = dev_data->tx_errors;
+			dev_data->state_chg_cb(dev, state, error_counters,
+					       dev_data->state_chg_cb_data);
```

Comment:
You need to make a copy of `dev_data->state_chg_cb`, check it against `NULL` and if non-NULL, call the copy. This code may lead to the `dev_data->state_chg_cb` value being changed between checking it and using it.